In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [24]:
df=pd.read_csv("Walmart_Store_sales.csv")

# Data Cleaning

In [25]:
df.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

In [26]:
df['Store'].nunique()

45

In [27]:
df.isnull().sum()

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


In [29]:
df['Date'] = pd.to_datetime(df['Date'])


In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         6435 non-null   int64         
 1   Date          6435 non-null   datetime64[ns]
 2   Weekly_Sales  6435 non-null   float64       
 3   Holiday_Flag  6435 non-null   int64         
 4   Temperature   6435 non-null   float64       
 5   Fuel_Price    6435 non-null   float64       
 6   CPI           6435 non-null   float64       
 7   Unemployment  6435 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(2)
memory usage: 402.3 KB


In [31]:
df.dtypes

Store                    int64
Date            datetime64[ns]
Weekly_Sales           float64
Holiday_Flag             int64
Temperature            float64
Fuel_Price             float64
CPI                    float64
Unemployment           float64
dtype: object

In [32]:
# Descriptive Statistics

In [35]:
df.describe().T

count          mean            std         min         25%  \
Store         6435.0  2.300000e+01      12.988182       1.000      12.000   
Weekly_Sales  6435.0  1.046965e+06  564366.622054  209986.250  553350.105   
Holiday_Flag  6435.0  6.993007e-02       0.255049       0.000       0.000   
Temperature   6435.0  6.066378e+01      18.444933      -2.060      47.460   
Fuel_Price    6435.0  3.358607e+00       0.459020       2.472       2.933   
CPI           6435.0  1.715784e+02      39.356712     126.064     131.735   
Unemployment  6435.0  7.999151e+00       1.875885       3.879       6.891   

                        50%           75%           max  
Store             23.000000  3.400000e+01  4.500000e+01  
Weekly_Sales  960746.040000  1.420159e+06  3.818686e+06  
Holiday_Flag       0.000000  0.000000e+00  1.000000e+00  
Temperature       62.670000  7.494000e+01  1.001400e+02  
Fuel_Price         3.445000  3.735000e+00  4.468000e+00  
CPI              182.616521  2.127433e+02  2.272328e+02  
Unemployment       7.874000  8.622000e+00  1.431300e+01

In [36]:
df.head()

Store       Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1 2010-05-02    1643690.90             0        42.31       2.572   
1      1 2010-12-02    1641957.44             1        38.51       2.548   
2      1 2010-02-19    1611968.17             0        39.93       2.514   
3      1 2010-02-26    1409727.59             0        46.63       2.561   
4      1 2010-05-03    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

# (1) Which store has maximum sales?

In [54]:
import pandas as pd

total_Sales = df.groupby('Store')['Weekly_Sales'].sum().round().sort_values(ascending=False)
total_Sales.head()




Store
20    301397792.0
4     299543953.0
14    288999911.0
13    286517704.0
2     275382441.0
Name: Weekly_Sales, dtype: float64

In [55]:
pd.DataFrame(total_sales).head(1)

Weekly_Sales
Store              
20      301397792.0

# Minimum_Sales

In [58]:
pd.DataFrame(total_sales).tail(1)

Weekly_Sales
Store              
33       37160222.0

# (2) Which store has maximum standard deviation i.e., the sales vary a lot. Also, find out¶
# the coefficient of mean to standard deviation

In [59]:
df_std = df.groupby('Store')['Weekly_Sales'].std().round(3).sort_values(ascending=False)

In [61]:
pd.DataFrame(df_std).head(1)

Weekly_Sales
Store              
14       317569.949

In [68]:
store14 = df[df.Store == 14].Weekly_Sales
store14

1859    2623469.95
1860    1704218.84
1861    2204556.70
1862    2095591.63
1863    2237544.75
           ...    
1997    1522512.20
1998    1687592.16
1999    1639585.61
2000    1590274.72
2001    1704357.62
Name: Weekly_Sales, Length: 143, dtype: float64

In [71]:
mean_to_stddev = store14.std() / store14.mean() * 100
mean_to_stddev = round(mean_to_stddev, 2)  # 
print(mean_to_stddev)




15.71


# (3) Which store/s has good quarterly growth rate in Q3’2012 ?

In [72]:
q2_sales = df[(df['Date'] >= '2012-04-01') & (df['Date'] <= '2012-06-30')].groupby('Store')['Weekly_Sales'].sum().round()
q3_sales = df[(df['Date'] >= '2012-07-01') & (df['Date'] <= '2012-09-30')].groupby('Store')['Weekly_Sales'].sum().round()

In [73]:
pd.DataFrame({'Q2 Sales':q2_sales,
              'Q3 Sales':q3_sales,
              'Difference':(q3_sales-q2_sales),
              'Growth Rate':(q3_sales-q2_sales)/q3_sales*100}).sort_values(by=['Growth Rate'], ascending=False).head()

Q2 Sales    Q3 Sales  Difference  Growth Rate
Store                                                 
16      6626133.0   6441311.0   -184822.0    -2.869323
7       7613594.0   7322394.0   -291200.0    -3.976841
35     10753571.0  10252123.0   -501448.0    -4.891163
26     13218290.0  12417575.0   -800715.0    -6.448240
39     20191586.0  18899955.0  -1291631.0    -6.834043

### No store has shown quarterly growth in quarter'3 2012.¶ 

# (4) Some holidays have a negative impact on sales. Find out holidays which have higher¶?

In [74]:
# Holiday Events

Super_Bowl = ['12-2-2010', '11-2-2011', '10-2-2012']
Labour_Day =  ['10-9-2010', '9-9-2011', '7-9-2012']
Thanksgiving =  ['26-11-2010', '25-11-2011', '23-11-2012']
Christmas = ['31-12-2010', '30-12-2011', '28-12-2012']

In [76]:
# Calculating holiday events sales
Super_Bowl_sales = round(df.loc[df['Date'].isin(Super_Bowl), 'Weekly_Sales'].mean(), 2)
Labour_Day_sales = round(df.loc[df['Date'].isin(Labour_Day), 'Weekly_Sales'].mean(), 2)
Thanksgiving_sales = round(df.loc[df['Date'].isin(Thanksgiving), 'Weekly_Sales'].mean(), 2)
Christmas_sales = round(df.loc[df['Date'].isin(Christmas), 'Weekly_Sales'].mean(), 2)


In [77]:
Super_Bowl_sales, Labour_Day_sales, Thanksgiving_sales, Christmas_sales

(1079127.99, 1042427.29, 1471273.43, 960833.11)

In [79]:
non_holiday_sales = round(df[(df['Holiday_Flag'] == 0)]['Weekly_Sales'].mean(),2)
non_holiday_sales

1041256.38

In [80]:
result = pd.DataFrame([{'Super Bowl Sales':Super_Bowl_sales,
              'Labour Day Sales':Labour_Day_sales,
              'Thanksgiving Sales':Thanksgiving_sales,
              'Christmas Sales':Christmas_sales,
              'Non Holiday Sales':non_holiday_sales}]).T

result

0
Super Bowl Sales    1079127.99
Labour Day Sales    1042427.29
Thanksgiving Sales  1471273.43
Christmas Sales      960833.11
Non Holiday Sales   1041256.38

### Thanksgiving has the highest sales (1,471,273.43) than non-holiday sales (1,041,256.38).¶

# (5) Provide a monthly and semester view of sales in units and give insights.

In [81]:
df['Day'] = pd.DatetimeIndex(df['Date']).day
df['Month'] = pd.DatetimeIndex(df['Date']).month
df['Year'] = pd.DatetimeIndex(df['Date']).year

In [82]:
df.head()

Store       Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1 2010-05-02    1643690.90             0        42.31       2.572   
1      1 2010-12-02    1641957.44             1        38.51       2.548   
2      1 2010-02-19    1611968.17             0        39.93       2.514   
3      1 2010-02-26    1409727.59             0        46.63       2.561   
4      1 2010-05-03    1554806.68             0        46.50       2.625   

          CPI  Unemployment  Day  Month  Year  
0  211.096358         8.106    2      5  2010  
1  211.242170         8.106    2     12  2010  
2  211.289143         8.106   19      2  2010  
3  211.319643         8.106   26      2  2010  
4  211.350143         8.106    3      5  2010